### PR TITLE
thuang-909-validate-csv

### PR DESCRIPTION
**Pull request by tihuan**
Live link:

https://thuang-frontend.rdev.single-cell.czi.technology/?gs=true

#### Reviewers
**Functional:** 
@MDunitz 

**Readability:** 
@Bento007 

---

## Changes
- add
1. Dependency: `jschardet` and `papaparse` for checkig CSV encoding and parsing CSV rows respectively
2. Add `parseGeneSetsCSV()` util function to parse a CSV file. This is temporarily used on homepage behind feature flag `gs=true` for us to test out different CSV files and also for E2E tests
3. `parseGeneSetsCSV()` is responsible for both validating CSV input and checking the encoding of the CSV file. The output is parsed and converted result that matches BE's payload format and error messages
4. `validateOutput()` is used by `parseGeneSetsCSV()` to validate the CSV requirements shown in [this epic](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/115). Each validator is encapsulated as a function. E.g., `hasGeneSetName()`, `hasGeneSetDescription()`, `hasGeneSymbol()`. The helper function also injects other CSV parsing errors reported by library `papaparse`
5. `frontend/tests/features/geneSets.test.ts` runs E2E tests on different CSV files

Example result using `/fixtures/geneSets/Tidy CSV With Comments, empty gene set name, empty gene set description, empty gene symbol, empty gene description, duplicate gene set name, and duplicate gene symbol UTF8.csv`:

Screen Shot 2021-03-18 at 2 11 42 PM

## Known Issues:
1. we need to **allow ASCII**, since If an `UTF-8` CSV file only has latin letters, it's the same as ASCII and `jschardet` will detect it as `ASCII` instead of `UTF-8`. [Source](https://www.azavea.com/blog/2014/03/24/solving-unicode-problems-in-python-2-7/#1_str_is_for_bytes_not_strings)

## Definition of Done (from ticket)
https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/909

- [x] Create test gene set files containing multiple gene sets representing the validation errors above
- [x] Test that all of the validation errors are caught and the error response specifies the gene set name and the validation error
- [x] All validation tests should be run each time (don't stop when one fails to ensure uploading gene sets isn't an endless nightmare of errors)

## QA steps (optional)

1. Pull branch
2. `npm i && npm start`
3. Visit: `http://localhost:8000/?gs=true`
4. Validate any CSV files in `frontend/tests/features/fixtures/geneSets/*` using the select file button
5. Verify the output is as expected
6. Repeat step 4 with other gene set CSV files you have and verify the output is as expected 🙏

### CODE REVIEW

File: frontend/tests/features/geneSets.test.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import { goToPage } from "tests/utils/helpers";
```
Comment (tihuan) [inline frontend/tests/features/geneSets.test.ts:1]: Different test cases!

**Review by tihuan — COMMENTED** [review]
Ready for review now 🙏 Thanks all!

File: frontend/src/views/Homepage/index.tsx [hunk 2]
```diff
@@ -1,8 +1,29 @@
+import loadable from "@loadable/component";
 import React, { FC } from "react";
+import { get } from "src/common/featureFlags";
+import { FEATURES } from "src/common/featureFlags/features";
+import { BOOLEAN } from "src/common/localStorage/set";
 import Collections from "src/components/Collections";
 
+const AsyncUploadCSV = loadable(
```
Comment (tihuan) [inline frontend/src/views/Homepage/index.tsx:8]: Code split `UploadCSV` component to only load if `?csv=true` in URL, since the component is very heavy!

**Comment by tihuan** [comment]
@brianraymor when get a chance, can you help play with this [live link](https://thuang-frontend.rdev.single-cell.czi.technology/?gs=true) and see if the validation is working as you expected?

Thank you!

**Comment by brianraymor** [comment]
Nice @tihuan - I appreciate the _live_ link for testing.

Success:

1. Tested variations of missing or misnamed headers
2. Tested missing gene set name
3. Tested duplicate gene set name (out of order)
3. Tested missing gene symbol in gene set
4. Tested duplicate gene symbol in gene set
5. Tested missing gene set description for first instance
5. Other variations of missing values

Failure and edge case but it's my fault:
1. I believe that our column header names should be **reserved**. For example, what does it mean if column A is 
`gene_set_name` and column F is also `gene_set_name`? Perhaps we don't care although it might confuse consumers. 

I did call out the names as **reserved** (more of SHOULD than a MUST) in my PR but did not indicate a subsequent validation failure. 

@hthomas-czi - any positions? 

It's good that the validation provides a complete list of errors, although a missing gene set name and symbol result in some ambiguity about gene set `''`

```
  "errorMessages": [
    "Missing Gene Set Name",
    "Gene set: '' is missing Gene Symbol"
  ]
```

Just curious @tihuan  - no action required beyond thought experiment - how hard would it be to report row numbers in the error message - "Row 2 Gene Description is missing" for easier "debugging" by curators ? @hthomas-czi - thoughts?

**Comment by hthomas-czi** [comment]
> Just curious @tihuan - no action required beyond thought experiment - how hard would it be to report row numbers in the error message - "Row 2 Gene Description is missing" for easier "debugging" by curators ? @hthomas-czi - thoughts?

I agree. I even included that concept in [the mockup](https://www.figma.com/file/TKV4sJ00zWobEFkaNgklC3/Gene-Sets?node-id=329%3A250) for missing gene set name, but we could take it even further as you suggested. 

Regarding your question about reserved columns, I don't really have an opinion.

**Comment by tihuan** [comment]
Thanks so much for the thorough testing and the discussion! 🎉 

@brianraymor Adding row number is a great idea! I think it's pretty trivial to add that. So every error message will just get their row number too?

E.g.,

`Row 4: Gene set: '${gene_set_name}' is missing Gene Set Description`

**Comment by tihuan** [comment]
If needed I can add the row number today!

File: frontend/src/common/utils/parseGeneSetsCSV/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  const [result] = await Promise.all([
+    validateOutputPromise,
+    checkEncodingPromise,
+  ]);
+
+  return result;
+}
+
+function createCheckEncoding({
```
[the remark this answers is not in the document]
Comment (brianraymor) [reply frontend/src/common/utils/parseGeneSetsCSV/index.ts:59]: Can you set the encoding rather than detecting the encoding ?

**Comment by brianraymor** [comment]
If the row number is _cheap_, go for it, @tihuan.

**Comment by tihuan** [comment]
OK!

**Comment by tihuan** [comment]
@brianraymor  Done 😎 👌 

Live link: https://thuang-frontend.rdev.single-cell.czi.technology/?gs=true

Screen Shot 2021-03-23 at 3 11 17 PM

**Comment by brianraymor** [comment]
It is off by 1 because the header row was not accounted for? For example:

`{"errorMessages":["Row 1: Gene set: 'club.cell' is missing Gene Set Description"],`

should be row 2 because row 1 is headers. I'm assuming that curators are editing in excel or google sheets and would be looking for errors on row 2.

**Comment by tihuan** [comment]
Ooh it's 0 based indexing! I forgot spreadsheets use 1 based indexing lol Will update!

**Comment by tihuan** [comment]
Updated!

Screen Shot 2021-03-23 at 3 40 50 PM

**Comment by brianraymor** [comment]
Confirmed.

**Comment by brianraymor** [comment]
Minor error - `"Row 3: Gene set symbol: 'ACK' already exists in the gene set: 'club.cell'"`

should be just Gene symbol ...

**Comment by tihuan** [comment]
Ah great catch thank you 🙏 ! Updated 😄 👍 

https://thuang-frontend.rdev.single-cell.czi.technology/

Screen Shot 2021-03-23 at 4 01 36 PM

File: frontend/jest/featureFlags.js [hunk 4]
```diff
@@ -7,6 +7,8 @@ module.exports = {
       localStorage: [
         { name: "cxg-ff-auth", value: "true" },
         { name: "cxg-ff-cc", value: "true" },
+        { name: "cxg-ff-gs", value: "true" },
```
[the remark this answers is not in the document]
Comment (MDunitz) [reply frontend/jest/featureFlags.js:10]: do these need to be changed to yes now?

**Review by MDunitz — APPROVED** [review]
lgtm with one small question about feature flags. Thanks!

File: frontend/jest/featureFlags.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tihuan) [reply frontend/jest/featureFlags.js:10]: Oh critical catch!! ⭐ 🙏 Thanks so much! Yes I'll change it here and the current ones in `main` too!

File: frontend/jest/featureFlags.js [hunk 5]
```diff
@@ -7,6 +7,7 @@ module.exports = {
       localStorage: [
         { name: "cxg-ff-auth", value: "yes" },
         { name: "cxg-ff-cc", value: "yes" },
+        { name: "cxg-ff-gs", value: "yes" },
```
Comment (tihuan) [inline frontend/jest/featureFlags.js:10]: @MDunitz Corrected the values in both `main` and this PR now 🙏 Thanks so much again for catching this!!

**Pull request merged**